README.md
# InSpec Compliance for the OpenStack Security Guide

This is a collection of [InSpec](http://inspec.io) scripts to check compliance against the [OpenStack Security Guide](http://docs.openstack.org/security-guide/).

The control checklists for Keystone, Horizon, Cinder, Nova and Neutron are implemented based on OpenStack Mitaka and beyond configuration standards.

Some control implementation exists for Swift and Manila, but has not been tested.

Beta-level controls exist for Glance. These controls are inspired by those currently recommended in the OpenStack Security Guide for Cinder.

## Installation

```shell
git clone git@github.com:chef-partners/inspec-openstack-security.git
cd inspec-openstack-security
bundle install
```

## Run tests locally

```shell
bundle exec inspec exec .
```

## Run tests against remote host(s)

Note that the controls can only be run against a single host until
https://github.com/chef/inspec/issues/268 is closed.

If your OpenStack control plane consists of multiple hosts, you'll need to
run InSpec against each host separately.

```shell
bundle exec inspec exec . -t ssh://user@hostname
```

## Run controls for a particular service

### Identity controls

```shell
bundle exec inspec exec . \
  --controls check-identity-01 check-identity-02 \
    check-identity-03 check-identity-04 \
    check-identity-05 check-identity-06
```

### Dashboard controls

```shell
bundle exec inspec exec . \
  --controls check-dashboard-01 check-dashboard-02 \
    check-dashboard-03 check-dashboard-04 \
    check-dashboard-05 check-dashboard-06 \
    check-dashboard-07 check-dashboard-08 \
    check-dashboard-09 check-dashboard-10 \
    check-dashboard-11
```

### Block Storage controls

```shell
bundle exec inspec exec . \
  --controls check-block-01 check-block-02 \
    check-block-03 check-block-04 \
    check-block-05 check-block-06 \
    check-block-07 check-block-08
```

### Compute controls

```shell
bundle exec inspec exec . \
  --controls check-compute-01 check-compute-02 \
    check-compute-03 check-compute-04 \
    check-compute-05
```

### Network controls

```shell
bundle exec inspec exec . \
  --controls check-neutron-01 check-neutron-02 \
    check-neutron-03 check-neutron-04 \
    check-neutron-05
```

### Image controls

```shell
bundle exec inspec exec . \
  --controls check-image-01 check-image-02 \
    check-image-03 check-image-04
```

### Messaging controls

```shell
bundle exec inspec exec . \
  --controls check-messaging-01 check-messaging-02 \
  check-messaging-03 check-messaging-04 \
  check-messaging-05 check-messaging-06 \
  check-messaging-07
```

### Orchestration controls
```shell
bundle exec inspec exec . \
  --controls check-orchestration-01 check-orchestration-02 \
    check-orchestration-03 --attrs attributes.yml
```

attributes.yml has the following contents
```yaml
heat_enabled: true
```

### Telemetry and Telemetry Alarming controls

```shell
inspec exec . --controls check-telemetry-01 check-telemetry-02 \
                check-telemetry-03 check-telemetry-04 \
                check-telemetry-alarming-01 check-telemetry-alarming-02 \
                check-telemetry-alarming-03 \
                --attrs attributes.yml
```

attributes.yml has the following contents
```yaml
ceilometer_enabled: true
aodh_enabled: true
```

# License

Apache 2

## License & Authors

- Author: JJ Asghar ([jj@chef.io](mailto:jj@chef.io))

```text
Copyright:: 2015-2017, Chef Software, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Changelog

## [1.0.4](https://github.com/dev-sec/openstack-baseline/tree/1.0.4) (2023-05-02)

[Full Changelog](https://github.com/dev-sec/openstack-baseline/compare/1.0.3...1.0.4)

**Implemented enhancements:**

- use centralised issue templates and workflows [\#33](https://github.com/dev-sec/openstack-baseline/pull/33) ([schurzi](https://github.com/schurzi))

**Merged pull requests:**

- add spellchecking with codespell [\#35](https://github.com/dev-sec/openstack-baseline/pull/35) ([schurzi](https://github.com/schurzi))
- Configure Renovate [\#34](https://github.com/dev-sec/openstack-baseline/pull/34) ([renovate[bot]](https://github.com/apps/renovate))
- Change linting to Cookstyle [\#32](https://github.com/dev-sec/openstack-baseline/pull/32) ([schurzi](https://github.com/schurzi))

## [1.0.3](https://github.com/dev-sec/openstack-baseline/tree/1.0.3) (2022-01-12)

[Full Changelog](https://github.com/dev-sec/openstack-baseline/compare/1.0.2...1.0.3)

**Merged pull requests:**

- use input instead of attribute [\#31](https://github.com/dev-sec/openstack-baseline/pull/31) ([micheelengronne](https://github.com/micheelengronne))
- fix rubocop error for Rakefile [\#29](https://github.com/dev-sec/openstack-baseline/pull/29) ([schurzi](https://github.com/schurzi))
- add dependency to chef-config for CI [\#28](https://github.com/dev-sec/openstack-baseline/pull/28) ([schurzi](https://github.com/schurzi))
- use version tag for changelog action [\#27](https://github.com/dev-sec/openstack-baseline/pull/27) ([schurzi](https://github.com/schurzi))
- Fix lint [\#26](https://github.com/dev-sec/openstack-baseline/pull/26) ([schurzi](https://github.com/schurzi))
- GitHub action [\#25](https://github.com/dev-sec/openstack-baseline/pull/25) ([rndmh3ro](https://github.com/rndmh3ro))

## [1.0.2](https://github.com/dev-sec/openstack-baseline/tree/1.0.2) (2020-07-23)

[Full Changelog](https://github.com/dev-sec/openstack-baseline/compare/1.0.1...1.0.2)

**Merged pull requests:**

- The release draft references the correct SHA [\#24](https://github.com/dev-sec/openstack-baseline/pull/24) ([micheelengronne](https://github.com/micheelengronne))

## [1.0.1](https://github.com/dev-sec/openstack-baseline/tree/1.0.1) (2020-06-18)

[Full Changelog](https://github.com/dev-sec/openstack-baseline/compare/1.0.0...1.0.1)

## [1.0.0](https://github.com/dev-sec/openstack-baseline/tree/1.0.0) (2020-06-18)

[Full Changelog](https://github.com/dev-sec/openstack-baseline/compare/0f1838390f4d01832fde46cbe9cb91eb6bb7e009...1.0.0)

**Closed issues:**

- No schema.json in glance [\#12](https://github.com/dev-sec/openstack-baseline/issues/12)

**Merged pull requests:**

- github actions release [\#23](https://github.com/dev-sec/openstack-baseline/pull/23) ([micheelengronne](https://github.com/micheelengronne))
- Update issue templates [\#22](https://github.com/dev-sec/openstack-baseline/pull/22) ([rndmh3ro](https://github.com/rndmh3ro))
- use recommended spdx license identifier [\#21](https://github.com/dev-sec/openstack-baseline/pull/21) ([chris-rock](https://github.com/chris-rock))
- update metadata [\#20](https://github.com/dev-sec/openstack-baseline/pull/20) ([chris-rock](https://github.com/chris-rock))
- Remove puts statement [\#19](https://github.com/dev-sec/openstack-baseline/pull/19) ([edtubillara](https://github.com/edtubillara))
- Lint project [\#17](https://github.com/dev-sec/openstack-baseline/pull/17) ([chris-rock](https://github.com/chris-rock))
- Refactored object inspec checks [\#16](https://github.com/dev-sec/openstack-baseline/pull/16) ([narengan](https://github.com/narengan))
-  converted swift config file owner to variable [\#15](https://github.com/dev-sec/openstack-baseline/pull/15) ([narengan](https://github.com/narengan))
- Implement Ceilometer and AODH controls based on Cinder [\#14](https://github.com/dev-sec/openstack-baseline/pull/14) ([edtubillara](https://github.com/edtubillara))
- Implement Heat controls based on Cinder [\#13](https://github.com/dev-sec/openstack-baseline/pull/13) ([edtubillara](https://github.com/edtubillara))
- Add attribute for glance config files [\#11](https://github.com/dev-sec/openstack-baseline/pull/11) ([edtubillara](https://github.com/edtubillara))
- Implement Glance controls based on Cinder guidance [\#10](https://github.com/dev-sec/openstack-baseline/pull/10) ([trumant](https://github.com/trumant))
- Removing duplicate check-dashboard-08 definition [\#9](https://github.com/dev-sec/openstack-baseline/pull/9) ([trumant](https://github.com/trumant))
- Fix cinder check typo [\#8](https://github.com/dev-sec/openstack-baseline/pull/8) ([trumant](https://github.com/trumant))
- Adding per-service controls examples for tested services [\#7](https://github.com/dev-sec/openstack-baseline/pull/7) ([trumant](https://github.com/trumant))
- Fixing keystone-paste filename [\#6](https://github.com/dev-sec/openstack-baseline/pull/6) ([trumant](https://github.com/trumant))
- Removing comments about Security guide being incorrect [\#5](https://github.com/dev-sec/openstack-baseline/pull/5) ([trumant](https://github.com/trumant))
- Complete refactor of Keystone, Cinder, Nova, Neutron and Horizon controls/checks [\#4](https://github.com/dev-sec/openstack-baseline/pull/4) ([trumant](https://github.com/trumant))
- Use new InSpec profile structure [\#1](https://github.com/dev-sec/openstack-baseline/pull/1) ([chris-rock](https://github.com/chris-rock))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


controls/check-block.rb
# encoding: utf-8
# All checks from http://docs.openstack.org/security-guide/block-storage/checklist.html

cinder_conf_dir = '/etc/cinder'
cinder_conf_file = "#{cinder_conf_dir}/cinder.conf"

control 'check-block-01' do
  title 'Cinder config files should be owned by root user and cinder group.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-01-is-user-group-ownership-of-config-files-set-to-root-cinder'

  describe file(cinder_conf_file) do
    it { should be_owned_by 'root' }
    its('group') { should eq 'cinder' }
  end
  describe file("#{cinder_conf_dir}/api-paste.ini") do
    it { should be_owned_by 'root' }
    its('group') { should eq 'cinder' }
  end
  describe file("#{cinder_conf_dir}/policy.json") do
    it { should be_owned_by 'root' }
    its('group') { should eq 'cinder' }
  end
  describe file("#{cinder_conf_dir}/rootwrap.conf") do
    it { should be_owned_by 'root' }
    its('group') { should eq 'cinder' }
  end
end

control 'check-block-02' do
  title 'Strict permissions should be set for all Cinder config files.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-02-are-strict-permissions-set-for-configuration-files'

  describe file(cinder_conf_file) do
    its('mode') { should cmp '0640' }
  end
  describe file("#{cinder_conf_dir}/api-paste.ini") do
    its('mode') { should cmp '0640' }
  end
  describe file("#{cinder_conf_dir}/policy.json") do
    its('mode') { should cmp '0640' }
  end
  describe file("#{cinder_conf_dir}/rootwrap.conf") do
    its('mode') { should cmp '0640' }
  end
end

control 'check-block-03' do
  title 'Cinder should use Keystone for authentication.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-03-is-keystone-used-for-authentication'

  # nil is acceptable as keystone is default value
  describe ini(cinder_conf_file) do
    its(['DEFAULT', 'auth_strategy']) { should be_nil.or eq 'keystone' }
  end
end

control 'check-block-04' do
  title 'Cinder should communicate with Keystone using TLS.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-04-is-tls-enabled-for-authentication'

  describe ini(cinder_conf_file) do
    its(['keystone_authtoken', 'auth_uri']) { should match(/^https:/) }
    # nil is acceptable as false is the default value
    its(['keystone_authtoken', 'insecure']) { should be_nil.or eq 'False' }
  end
end

control 'check-block-05' do
  title 'Cinder should communicate with Nova using TLS.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-05-does-cinder-communicate-with-nova-over-tls'

  # nil is acceptable as false is the default value
  describe ini(cinder_conf_file) do
    its(['DEFAULT', 'nova_api_insecure']) { should be_nil.or eq 'False' }
  end
end

control 'check-block-06' do
  title 'Cinder should communicate with Glance using TLS.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-06-does-cinder-communicate-with-glance-over-tls'

  describe ini(cinder_conf_file) do
    # nil is acceptable as the glance endpoint may be sourced from Keystone based on the value of glance_catalog_info
    its(['DEFAULT', 'glance_api_servers']) { should be_nil.or match(/^https:/) }
    # nil is acceptable as false is the default value
    its(['DEFAULT', 'glance_api_insecure']) { should be_nil.or eq 'False' }
  end
end

control 'check-block-07' do
  title 'Cinder should use secure NAS permissions.'
  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-07-is-nas-operating-in-a-secure-environment'

  cinder_conf = ini(cinder_conf_file)

  only_if do
    cinder_conf.value(['DEFAULT', 'nas_host']).nil?
  end

  describe ini(cinder_conf_file) do
    # nil is acceptable as auto is the default value
    its(['DEFAULT', 'nas_secure_file_permissions']) { should be_nil.or eq('True').or eq('auto') }
    # nil is acceptable as auto is the default value
    its(['DEFAULT', 'nas_secure_file_operations']) { should be_nil.or eq('True').or eq('auto') }
  end
end

control 'check-block-08' do

  title 'Max request body size should be set to avoid denial of service attacks.'

  ref 'http://docs.openstack.org/security-guide/block-storage/checklist.html#check-block-08-is-max-size-for-the-body-of-a-request-set-to-default-114688'

  describe ini(cinder_conf_file) do
    # nil is acceptable as 114688 is the default value
    its(['DEFAULT', 'osapi_max_request_body_size']) { should be_nil.or be >= 114688 }
  end
end
